File: 09.algos-star-wars/episode01/skywalkerPower.ts
export type Jedi = {
    name: string
    power: number
    children: Jedi[]
}

export const skywalkerPowerWithFor = (root: Jedi): number => {
    if (!root) return 0;

    let total = root.power;
    for (const child of root.children) {
        total += skywalkerPowerWithFor(child);
    }
    return total;
};

const skywalkerPowerForPadawans = (jedis: Jedi[]) => jedis.reduce(
    (power, child) => power + skywalkerPower(child),
    0);

export const skywalkerPower = (root: Jedi): number => {
    if (!root) return 0;
    return root.power + skywalkerPowerForPadawans(root.children);
};

File: 09.algos-star-wars/episode01/java/src/main/java/skywalker/SkywalkerPower.java
package skywalker;

import java.util.List;

public class SkywalkerPower {

    public record Jedi(String name, int power, List<Jedi> children) {}

    public static int skywalkerPower(Jedi root) {
        // TODO
        return 0;
    }
}
